main/test/ca/uwaterloo/flix/api/lsp/TestCompletionProvider.scala

```diff
@@ -318,17 +339,113 @@ class TestCompletionProvider extends AnyFunSuite {
+      }
+    }
+    Assert.Ok
+  }
+
+  /**
+    * Returns all *block comment* tokens in the given program `prg` associated with the given AST `root`.
+    *
+    * That is, comments surrounded by `/* */`.
+    */
+  private def blockCommentsOf(prg: String, root: Root): List[Token] = {
+    def isBlockComment(tok: Token): Boolean = tok.kind match {
+      case TokenKind.CommentBlock => true
+      case _ => false
+    }
+
+    getTokens(prg, root).filter(isBlockComment)
+  }
+
+  /**
+    * Returns all *line comment* and *doc comment* tokens in the given program `prg` associated with the given AST `root`.
+    *
+    * Returns both regular line comments and documentation comments. That is, comments starting with `//` or `///`.
+    */
+  private def lineCommentsOf(prg: String, root: Root): List[Token] = {
+    def isLineOrDocComment(tok: Token): Boolean = tok.kind match {
+      case TokenKind.CommentLine => true
+      case TokenKind.CommentDoc => true
+      case _ => false
+    }
+
+    getTokens(prg, root).filter(isLineOrDocComment)
+  }
+
+  /**
+    * Returns all *keyword* tokens in the given program `prg` associated with the given AST `root`.
+    */
+  private def keywordsOf(prg: String, root: Root): List[Token] =
+    getTokens(prg, root).filter(_.kind.isKeyword)
+
+  /**
+    * Returns all *literal* tokens in the given program `prg` associated with the given AST `root`.
+    */
+  private def literalsOf(prg: String, root: Root): List[Token] =
+    getTokens(prg, root).filter(_.kind.isLiteral)
+
+
   /**
-    * Asserts that the given completion list is empty at the given position.
+    * Returns all tokens in the given program `prg` associated with the given AST `root`.
     */
-  private def assertEmpty(completions: List[CompletionItem], sourceLocation: SourceLocation, pos: Position): Unit = {
-    if (completions.nonEmpty) {
-      println(code(sourceLocation, s"Unexpected completions at $pos"))
-      println(s"Found completions: ${completions.map(_.label)}")
-      fail(s"Expected no completions at position $pos, but found ${completions.length} completions.")
+  private def getTokens(prg: String, root: Root): List[Token] =
+    // TODO: Can we get rid of the need for mkSource?
+    root.tokens(mkSource(prg)).toList
+
+  sealed trait Assert
+
+  private object Assert {
+
+    case object Ok extends Assert
+
+    case object Fail extends Assert
+
+    /**
+      * Returns `Assert.Ok` if `c` is `true`. Returns `Assert.Fail` otherwise.
+      */
+    def cond(c: Boolean): Assert = if (c) Ok else Fail
+
+    /**
+      * Asserts that the given list of completion `l` for position `pos` is empty.
+      */
+    def isEmpty(l: List[Completion], pos: Position): Assert = {
+      if (l.isEmpty) {
+        Assert.Ok
+      } else {
+        println(s"Found completions: ${l.map(_.toCompletionItem(Flix)).map(_.label)}")
+        fail(s"Expected no completions at position $pos, but found ${l.length} completions.")
```

Comment:
I don't understand this design; it seems to be mixing monad and side-effects. We return Ok if success, but throw an error if fail.

Reply:
I did not write that code. This is a partial refactor towards something more functional.